tls/s2n_post_handshake.c

```diff
@@ -13,66 +13,159 @@
-                break;
-            case TLS_CLIENT_HELLO:
-            case TLS_SERVER_HELLO:
-            case TLS_END_OF_EARLY_DATA:
-            case TLS_ENCRYPTED_EXTENSIONS:
-            case TLS_CERTIFICATE:
-            case TLS_SERVER_KEY:
-            case TLS_CERT_REQ:
-            case TLS_SERVER_HELLO_DONE:
-            case TLS_CERT_VERIFY:
-            case TLS_CLIENT_KEY:
-            case TLS_FINISHED:
-            case TLS_SERVER_CERT_STATUS:
-                /* All other known handshake messages should be rejected */
-                POSIX_BAIL(S2N_ERR_BAD_MESSAGE);
-                break;
-            default:
-                /* Ignore all other messages */
-                break;
+    return S2N_RESULT_OK;
+}
+
+/*
+ * Read a handshake message from conn->in.
+ *
+ * Handshake messages can be fragmented, meaning that a single message
+ * may be split between multiple records. conn->in only holds a single
+ * record at a time, so we may need to call this method multiple
+ * times to construct the complete message. We store the partial message
+ * in conn->post_handshake.in between calls.
+ */
+S2N_RESULT s2n_post_handshake_message_recv(struct s2n_connection *conn)
+{
+    RESULT_ENSURE_REF(conn);
+
+    struct s2n_stuffer *in = &conn->in;
+    struct s2n_stuffer *message = &conn->post_handshake.in;
+    uint8_t message_type = 0;
+    uint32_t message_len = 0;
+
+    /* We always start reading from the beginning of the message.
+     * Reset the read progress, but keep the write progress since we
+     * may already have read a partial message.
+     */
+    RESULT_GUARD_POSIX(s2n_stuffer_reread(message));
+
+    /* At minimum, the message stuffer needs to have enough space to read the header.
+     * For small messages like KeyUpdate and HelloRequest, this is all the space we will need.
+     */
+    if (s2n_stuffer_is_freed(message)) {
+        struct s2n_blob b = { 0 };
+        RESULT_GUARD_POSIX(s2n_blob_init(&b, conn->post_handshake.header_in,
+                sizeof(conn->post_handshake.header_in)));
+        RESULT_GUARD_POSIX(s2n_stuffer_init(message, &b));
+    }
+
+    /* Try to copy the header into the message stuffer.
+     * The message stuffer may already contain some or all of the header if
+     * we have read fragments of this message from previous records.
+     */
+    if (s2n_stuffer_data_available(message) < TLS_HANDSHAKE_HEADER_LENGTH) {
+        uint32_t remaining = TLS_HANDSHAKE_HEADER_LENGTH - s2n_stuffer_data_available(message);
+        uint32_t to_read = MIN(remaining, s2n_stuffer_data_available(in));
+        RESULT_GUARD_POSIX(s2n_stuffer_copy(in, message, to_read));
+    }
+    RESULT_ENSURE(s2n_stuffer_data_available(message) >= TLS_HANDSHAKE_HEADER_LENGTH, S2N_ERR_IO_BLOCKED);
+
+    /* Parse the header */
+    RESULT_GUARD(s2n_handshake_parse_header(message, &message_type, &message_len));
+    RESULT_ENSURE(message_len == 0 || s2n_stuffer_data_available(in), S2N_ERR_IO_BLOCKED);
```

Comment:
Why would the message length ever be 0? The header was one fragment and the message was another?

Reply:
It's unrelated to fragmentation, since this is the handshake message length, not the record length. Some handshake messages are empty, like HelloRequest.